File: app/src/main/java/net/tjado/usbgadget/MainActivity.java
package net.tjado.usbgadget;

import android.content.DialogInterface;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.RecyclerView;
import androidx.viewpager2.widget.ViewPager2;

import com.google.android.material.tabs.TabLayout;
import com.google.android.material.tabs.TabLayoutMediator;

import java.lang.reflect.Field;

public class MainActivity extends AppCompatActivity {

    private GadgetViewModel gadgetViewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        ViewPager2 viewPager2 = findViewById(R.id.viewpager);
        viewPager2.setAdapter(new ViewPager2FragmentStateAdapter(this));

        // reduce sensitivity of ViewPager2
        // https://stackoverflow.com/a/60672891
        try {
            Field recyclerViewField = ViewPager2.class.getDeclaredField("mRecyclerView");
            recyclerViewField.setAccessible(true);

            RecyclerView recyclerView = (RecyclerView) recyclerViewField.get(viewPager2);

            Field touchSlopField = RecyclerView.class.getDeclaredField("mTouchSlop");
            touchSlopField.setAccessible(true);

            int touchSlop = (int) touchSlopField.get(recyclerView);
            touchSlopField.set(recyclerView, touchSlop * 3);

        } catch (NoSuchFieldException | IllegalAccessException e) {
            e.printStackTrace();
        }

        TabLayout tabLayout = findViewById(R.id.tabs);
        TabLayoutMediator tabLayoutMediator = new TabLayoutMediator(tabLayout, viewPager2,
                new TabLayoutMediator.TabConfigurationStrategy() {
                    @Override
                    public void onConfigureTab(@NonNull TabLayout.Tab tab, int position) {
                        switch (position) {
                            case (0):
                                tab.setText("Gadgets");
                                break;
                            case (1):
                                tab.setText("Device Info");
                                break;
                            case (2):
                                tab.setText("Logs");
                                break;
                        }
                    }
                });
        tabLayoutMediator.attach();

        gadgetViewModel = new ViewModelProvider(this).get(GadgetViewModel.class);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();

        switch (id) {
            case R.id.action_refresh:
                refreshGadgets();
                return true;
            case R.id.action_add:
                addGadget();
                break;
            case R.id.info:
                showInfo();
                return true;
        }

        return super.onOptionsItemSelected(item);
    }

    protected void refreshGadgets() {
        gadgetViewModel.updateGadgetData();
    }

    protected void addGadget() {
        final AlertDialog.Builder alertBuilder = new AlertDialog.Builder(this);
        alertBuilder.setTitle("Add Gadget");
        alertBuilder.setCancelable(true);

        alertBuilder.setItems(gadgetViewModel.gadgetProfileList, (dialog, which) -> gadgetViewModel.loadGadgetProfile(gadgetViewModel.gadgetProfileList[which]));

        alertBuilder.show();
    }

    protected void showInfo() {
        final AlertDialog.Builder alertBuilder = new AlertDialog.Builder(this);
        alertBuilder.setTitle(R.string.info_title);
        alertBuilder.setCancelable(false);

        alertBuilder.setMessage(R.string.info_message);
        alertBuilder.setPositiveButton("Ok", new DialogInterface.OnClickListener() {
            public void onClick(DialogInterface dialoginterface, int i) {
            }});

        alertBuilder.show();
    }

}

File: app/src/main/java/net/tjado/usbgadget/MainActivity.kt
package net.tjado.usbgadget

import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.RecyclerView
import androidx.viewpager2.widget.ViewPager2
import com.google.android.material.tabs.TabLayoutMediator
import net.tjado.usbgadget.databinding.ActivityMainBinding


class MainActivity : AppCompatActivity() {
    private lateinit var binding: ActivityMainBinding

    private lateinit var gadgetViewModel: GadgetViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        binding.viewpager.adapter = ViewPager2FragmentStateAdapter(this)

        // reduce sensitivity of ViewPager2
        // https://stackoverflow.com/a/60672891
        try {
            val recyclerViewField = ViewPager2::class.java.getDeclaredField("mRecyclerView")
            recyclerViewField.isAccessible = true
            val recyclerView = recyclerViewField[binding.viewpager] as RecyclerView
            val touchSlopField = RecyclerView::class.java.getDeclaredField("mTouchSlop")
            touchSlopField.isAccessible = true
            val touchSlop = touchSlopField[recyclerView] as Int
            touchSlopField[recyclerView] = touchSlop * 3
        } catch (e: NoSuchFieldException) {
            e.printStackTrace()
        } catch (e: IllegalAccessException) {
            e.printStackTrace()
        }

        val tabLayoutMediator = TabLayoutMediator(
            binding.tabs, binding.viewpager, true, false
        ) { tab, position ->
            when (position) {
                0 -> tab.text = "Gadgets"
                1 -> tab.text = "Device Info"
                2 -> tab.text = "Gadget Manager"
                3 -> tab.text = "Logs"
            }
        }
        tabLayoutMediator.attach()

        gadgetViewModel = ViewModelProvider(this).get(GadgetViewModel::class.java)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_add ->  {
                addGadget()
                return true
            }
            R.id.action_refresh -> {
                refreshGadgets()
                return true
            }
            R.id.info -> {
                showInfo()
                return true
            }
        }

        return super.onOptionsItemSelected(item)
    }

    private fun refreshGadgets() {
        gadgetViewModel.updateGadgetData()
    }

    private fun addGadget() {
        AlertDialog.Builder(this)
            .setTitle("Add Gadget")
            .setCancelable(true)
            .setItems(gadgetViewModel.gadgetProfileList) { _, which ->
                gadgetViewModel.loadGadgetProfile(
                    gadgetViewModel.gadgetProfileList[which]
                )
            }.show()
    }

    private fun showInfo() {
        AlertDialog.Builder(this)
            .setTitle(R.string.info_title)
            .setCancelable(false)
            .setMessage(R.string.info_message)
            .setPositiveButton("Ok", null)
            .show()
    }

    fun setTabStatus(id: Int, isClickable: Boolean) {
       binding.tabs.getTabAt(id)?.view?.isClickable = isClickable
    }
}